pkg/volume/util/subpath/subpath_linux.go

```diff
@@ -587,6 +587,17 @@ func doSafeOpen(pathname string, base string) (int, error) {
 			return -1, fmt.Errorf("unexpected symlink found %s", currentPath)
 		}
 
+		// make sure to open directory on child filesystem such as bind mounted directory
+		if fileFmt == syscall.S_IFDIR {
+			if err = syscall.Close(childFD); err != nil {
+				return -1, fmt.Errorf("closing dir fd for %q failed: %v", currentPath, err)
+			}
+
+			if childFD, err = syscall.Openat(parentFD, seg+"/", openFDFlags|unix.O_CLOEXEC, 0); err != nil {
```

Comment:
I don't know the details, but just comparing with the previous function, the only difference is seg vs seg+"/"? 
Is it ok we always use seg+"/"? 

```
childFD, err = syscall.Openat(parentFD, seg, openFDFlags|unix.O_CLOEXEC, 0)
```

Reply:
As the commit log and the corresponding issue comment said,  if "child-dir" is a mount point for another file system, `openat("parent-dir", "child-dir")` actually open directory on file system "parent-dir", not the expected directory on file system "parent-dir/child-dir".

I got three versions of this patch:

(1) always open `seg+"/"`.

This is wrong, it breaks plain file bind mount.

(2) first try `seg+"/"`, if gets ENOTDIR error, try `seg` again.

This leads to security vulnerability and breaks Kubernetes CI tests, when `seg` is a symlink, `seg+"/"` will follow it, this behaviour conflicts with `O_NOFOLLOW` in `openFDFlags`.

(3) the current version,  first try `seg`, if it's a directory, try `seg+"/"` again to force traversing into a possible mount point.